README.md
# CS410 - Text Information Systems - Sentiment Analysis on Trending News

## Group Name - MPDL

## Members
1. Michaela Horn - mmhorn2@illinois.edu
2. Peter Jefferson - peterwj2@illinois.edu
3. Da'Mon Morris - dmorri25@illinois.edu
4. Lookman Olowo - lolowo2@illinois.edu


## Papers:
1. Project Proposal - [Project Proposal - MPDL - CS410.pdf](https://github.com/PeterWJefferson/CourseProject/blob/main/.docs/Project%20Proposal%20-%20MPDL%20-%20CS410.pdf)
    - Overview of the project
    - Team member contributions
2. Progress Report - [Progress Report - MPDL - CS410 Project.pdf](https://github.com/PeterWJefferson/CourseProject/blob/main/.docs/Progress%20Report%20-%20MPDL%20-%20CS410%20Project.pdf)
    
    - Progress made by Nov 14, 2022
    - Remaining tasks
    - Any challenges/issues being faced
3. Final Report - [Final Report - MPDL - CS410 Project.pdf](https://github.com/PeterWJefferson/CourseProject/blob/main/.docs/Final%20Report%20-%20MPDL%20-%20CS410%20Project.pdf)
    - Libraries used
    - Implementation details
4. Final Presentation - [CS410FinalProject.mp4](https://mediaspace.illinois.edu/media/t/1_m17gu9wt), [CS410FinalProject on YouTube](https://www.youtube.com/watch?v=nXy6tBeNWLU)
    - High-level overview of what our app is doing 
    - Diagram review
    - Review of requirements
    - Demonstration of application at work


## Test the application live
- Live Test Instance: [http://mpdl.pythonanywhere.com/](http://mpdl.pythonanywhere.com/)

## Running the application locally
- Download the code from github
- Prepare environment
  - Switch to environment using python version >3.6.0
  - Install the required dependencies: `pip install -r requirements.txt --upgrade`
  - Export the FLASK_APP environment variable to the app.py file: `export FLASK_APP=app.py`
- Start application
  - Execute flask: `flask run`
  - Open browser and go to the url http://localhost:5000


.semantics/CHANGELOG.md
# 1.0.0 (2022-12-08)


### Bug Fixes

* add utf-8 encoding when opening text file. ([6f6f986](https://github.com/PeterWJefferson/CourseProject/commit/6f6f986ac41991e7e4a35ff59d31a1c8c91947da))
* cleans up code, updates docs ([10240d8](https://github.com/PeterWJefferson/CourseProject/commit/10240d876fa616f142f47460abb6810f60e1fbcc))
* Correted the news source names from checkbox that gets sent into API since the syntax was off for some of the news orgs with spaces in them. i.e. @FoxNews was being input as fox_news. ([2aa56db](https://github.com/PeterWJefferson/CourseProject/commit/2aa56db22511e97f8a2e5103dbf9cd2d0c604ca9))
* overfitting bug removed ([a07db64](https://github.com/PeterWJefferson/CourseProject/commit/a07db64b0312c830e85fdf4fdd5de5d1b3198acf))
* re-cleans tweets after filtering search input ([8abd210](https://github.com/PeterWJefferson/CourseProject/commit/8abd210943bddfa1f31b5fb41293e782b364c1df))
* Removed upvote button/logic ([dc5e856](https://github.com/PeterWJefferson/CourseProject/commit/dc5e8560e59c49a8d18ddfeb77f6d9a30a191d60))


### Features

* Add Progress Report ([f012e45](https://github.com/PeterWJefferson/CourseProject/commit/f012e45b1e2ea77420dca7b3daf2e2c17b2c2c55))
* added delete lines functionality to delete all but 31 documents from the corpus when the "Click to delete previous searches" button is clicked on the home page. ([e2e3bd8](https://github.com/PeterWJefferson/CourseProject/commit/e2e3bd84fb34e96dc4b4dddeab1111960863c89d))
* Added delete recent search topic button to index.html, and back button and homepage UIUC logo to results.html. Created delete_search_topic() function to hold functionality to delete all the recent topics stored. ([f1f9f0a](https://github.com/PeterWJefferson/CourseProject/commit/f1f9f0a1a2a3f11e4c0a8f49b58578c57214df4a))
* added hyperlinking to each tweet result ([730334b](https://github.com/PeterWJefferson/CourseProject/commit/730334b79986cec9f5c5aefbc814f0a402567c5a))
* Added logic to route returned tweets to the results.html page. ([6c784f7](https://github.com/PeterWJefferson/CourseProject/commit/6c784f7144128b35d951b9e1d5fbceb243971106))
* Added new CSS styling ([8a229ef](https://github.com/PeterWJefferson/CourseProject/commit/8a229ef3fb7f8cee6a94fefd8ab95e0e278eb747))
* Added news source check boxes to search page and linked them to the backend to search those news sources only for the query if any of the boxes are selected. ([eca7977](https://github.com/PeterWJefferson/CourseProject/commit/eca797759a10e6e5749016981778ccf6789043bb))
* added percentage printout on results page of each sentiment by news source selected. ([2135947](https://github.com/PeterWJefferson/CourseProject/commit/2135947a93b9835981d0d25ba669458b734e4f89))
* added recent corpus and filter recent searches in corpus ([04d125f](https://github.com/PeterWJefferson/CourseProject/commit/04d125f8d636b94658ad53672bd38f6b6d55736e))
* added Recommended search topic buttons that populate the search bar when clicked. Also added recommended search skeleton for backend to use PLSA. Added title header and UIUC Logo. ([3cac386](https://github.com/PeterWJefferson/CourseProject/commit/3cac386b7a1371ce93bf7edcc21a803348d19108))
* added Recommended search topic buttons that populate the search bar when clicked. Also added recommended search skeleton for backend to use PLSA. Added title header and UIUC Logo. ([f210a48](https://github.com/PeterWJefferson/CourseProject/commit/f210a488dc3c42c930bdda9703098bfc2bdf371a))
* Added upvote button to each tweet row and logic to call the upvote_post() function on backend. ([2369342](https://github.com/PeterWJefferson/CourseProject/commit/2369342de92f4954e10e49b4f56405452dab0bd2))
* adds range to subjectivity score ([8f5044c](https://github.com/PeterWJefferson/CourseProject/commit/8f5044cf899388abafa166b8e0a048524b59a717))
* adds ranker ([dbb1560](https://github.com/PeterWJefferson/CourseProject/commit/dbb1560026587724c9bfee168764ec7d4232561b))
* adds source to response ([4ceb9ae](https://github.com/PeterWJefferson/CourseProject/commit/4ceb9aeed89fd0043d4c8d2fa84cc4b3651ef4b8))
* adds tweets to persistent corpus and results for q='Thanksgiving' ([d9ae149](https://github.com/PeterWJefferson/CourseProject/commit/d9ae14948984d35c4ee60b073a1de55a206f1d7b))
* changes to full-length tweets ([fb2c0df](https://github.com/PeterWJefferson/CourseProject/commit/fb2c0df22be7382d4978b16a6b477c628598e41b))
* changes to full-length tweets ([01fb3ce](https://github.com/PeterWJefferson/CourseProject/commit/01fb3ce1f84e66d9aa83fd23e352fa8c8dfa6e65))
* Cleaned up news source sentiment percetage print out on results page and added total number of tweets analyzied for sentiment analysis in printout. ([7407421](https://github.com/PeterWJefferson/CourseProject/commit/7407421298c6db3963c9cebb6d17ca8c0552c562))
* connects the API to the Twitter client functions ([c955f56](https://github.com/PeterWJefferson/CourseProject/commit/c955f56fd2f2c52dadb17474dcaa18a400e998e3))
* Created initial frontend files, added search bar and POST button that stores user query and jumps to the results page. Re-ordered file structure to get Flask frontend files to reference correctly via flask documentation: https://flask.palletsprojects.com/en/2.2.x/quickstart/#static-files ([7f861d1](https://github.com/PeterWJefferson/CourseProject/commit/7f861d1db9c8500222f356218d92696893f8dc21))
* Hyperlinked the Twitter accounts in the Account columns ([c24085d](https://github.com/PeterWJefferson/CourseProject/commit/c24085dbf66e4a6251c15191d671a69c714cd4e8))
* **news:** news tweets function works in api.py, had trouble with app.py ([f354776](https://github.com/PeterWJefferson/CourseProject/commit/f35477614f55fe01da4361c0696760f6a0f81201))
* **news:** received news tweets from query instead of any tweets ([2d8faec](https://github.com/PeterWJefferson/CourseProject/commit/2d8faec218418744307183d9467c417656c289a1))

Plsa.py
import numpy as np
import math
import heapq
import random

def normalize(input_matrix):
    """
    Normalizes the rows of a 2d input_matrix so they sum to 1
    """
    row_sums = input_matrix.sum(axis=1)
    try:
        assert (np.count_nonzero(row_sums)==np.shape(row_sums)[0]) # no row should sum to zero
    except Exception:
        raise Exception("Error while normalizing. Row(s) sum to zero")
    new_matrix = input_matrix / row_sums[:, np.newaxis]
    return new_matrix
       
class Corpus(object):
    """
    A collection of documents.
    """
    def __init__(self, documents_path):
        """
        Initialize empty document list.
        """
        self.documents = []
        self.vocabulary = []
        self.likelihoods = []
        self.documents_path = documents_path
        self.term_doc_matrix = None 
        self.document_topic_prob = None  # P(z | d)
        self.topic_word_prob = None  # P(w | z)
        self.topic_prob = None  # P(z | d, w)
        self.number_of_documents = 0
        self.vocabulary_size = 0
    def build_corpus(self):
        """
        Read document, fill in self.documents, a list of list of word
        self.documents = [["the", "day", "is", "nice", "the", ...], [], []...]
        
        Update self.number_of_documents
        """
        
        N_lines = 30
        with open(self.documents_path) as corpus_file:
            first_30 = [next(corpus_file) for x in range(N_lines)]
        for line in first_30:
            doc = list()
            doc.extend(line.split())
            self.documents.append(doc)
            self.number_of_documents += 1
                
        # with open(self.documents_path, 'r') as file:
        #     for line in file.readlines():
        #         doc = list()
        #         doc.extend(line.split())
        #         self.documents.append(doc)
        #         self.number_of_documents += 1

    def build_vocabulary(self):
        """
        Construct a list of unique words in the whole corpus. Put it in self.vocabulary
        for example: ["rain", "the", ...]
        Update self.vocabulary_size
        """
        _set = set()
        for document in self.documents:
            _set.update(document)
        self.vocabulary = _set
        self.vocabulary_size = len(self.vocabulary)
    def build_term_doc_matrix(self):
        """
        Construct the term-document matrix where each row represents a document, 
        and each column represents a vocabulary term.
        self.term_doc_matrix[i][j] is the count of term j in document i
        """
        
        self.term_doc_matrix = np.zeros(shape=(self.number_of_documents, self.vocabulary_size))
        vocab_dist = {j: i for i, j in enumerate(self.vocabulary)}
        for i, document in enumerate(self.documents):
            for t in document:
                self.term_doc_matrix[i][vocab_dist[t]] += 1

    def initialize_randomly(self, number_of_topics):
        """
        Randomly initialize the matrices: document_topic_prob and topic_word_prob
        which hold the probability distributions for P(z | d) and P(w | z): self.document_topic_prob, and self.topic_word_prob
        Don't forget to normalize! 
        HINT: you will find numpy's random matrix useful [https://docs.scipy.org/doc/numpy-1.15.0/reference/generated/numpy.random.random.html]
        """
        self.document_topic_prob = normalize(np.random.random(size=(self.number_of_documents, number_of_topics)))
        self.topic_word_prob = normalize(np.random.random(size=(number_of_topics, len(self.vocabulary))))
    def initialize_uniformly(self, number_of_topics):
        """
        Initializes the matrices: self.document_topic_prob and self.topic_word_prob with a uniform 
        probability distribution. This is used for testing purposes.
        DO NOT CHANGE THIS FUNCTION
        """
        self.document_topic_prob = np.ones((self.number_of_documents, number_of_topics))
        self.document_topic_prob = normalize(self.document_topic_prob)
        self.topic_word_prob = np.ones((number_of_topics, len(self.vocabulary)))
        self.topic_word_prob = normalize(self.topic_word_prob)
    def initialize(self, number_of_topics, random=False):
        """ Call the functions to initialize the matrices document_topic_prob and topic_word_prob
        """
        print("Initializing PLSA...")
        if random:
            self.initialize_randomly(number_of_topics)
        else:
            self.initialize_uniformly(number_of_topics)
    def expectation_step(self):
        """ The E-step updates P(z | w, d)
        """
        # print("E step:")
        
        self.topic_word_prob = np.nan_to_num(self.topic_word_prob)
        for document in range(self.topic_prob.shape[0]):
            for voc in range(self.topic_prob.shape[2]):
                self.topic_prob[document, :, voc] = self.document_topic_prob[document, :] * self.topic_word_prob[:, voc]
                self.topic_prob[document, :, voc] /= self.topic_prob[document, :, voc].sum()
        self.topic_word_prob = np.nan_to_num(self.topic_word_prob)
            
    def maximization_step(self, number_of_topics):
        """ The M-step updates P(w | z)
        """
        # print("M step:")
        
        # update P(w | z)
        
        for t in range(self.topic_prob.shape[1]):
            for voc in range(self.topic_prob.shape[2]):
                self.topic_word_prob[t, voc] = self.term_doc_matrix[:, voc].dot(self.topic_prob[:, t, voc])
            self.topic_word_prob[t, :] /= self.topic_word_prob[t, :].sum()
        self.topic_word_prob = np.nan_to_num(self.topic_word_prob)
        
        # update P(z | d)
        for document in range(self.topic_prob.shape[0]):
            for topic in range(self.topic_prob.shape[1]):
                self.document_topic_prob[document, topic] = self.term_doc_matrix[document, :].dot(self.topic_prob[document, topic, :])
            self.document_topic_prob[document, :] /= self.document_topic_prob[document, :].sum()
        self.document_topic_prob = np.nan_to_num(self.document_topic_prob)

    def calculate_likelihood(self, number_of_topics):
        """ Calculate the current log-likelihood of the model using
        the model's updated probability matrices
        
        Append the calculated log-likelihood to self.likelihoods
        """
        self.likelihoods.append(np.sum(np.log(np.matmul(self.document_topic_prob, self.topic_word_prob)) * self.term_doc_matrix))
        return self.likelihoods[-1]
    def plsa(self, number_of_topics, max_iter, epsilon):
        """
        Model topics.
        """
        # print ("EM iteration begins...")
        
        # build term-doc matrix
        self.build_term_doc_matrix()
        
        # Create the counter arrays.
        
        # P(z | d, w)
        self.topic_prob = np.zeros([self.number_of_documents, number_of_topics, self.vocabulary_size], dtype=np.float)
        # P(z | d) P(w | z)
        self.initialize(number_of_topics, random=False)
        # Run the EM algorithm
        current_likelihood = 0.0
        prev_prob = self.topic_prob.copy()
        for iteration in range(max_iter):
            self.expectation_step()
            prev_prob = self.topic_prob.copy()
            self.maximization_step(number_of_topics)
            self.calculate_likelihood(number_of_topics)
            tmp_likelihood = self.calculate_likelihood(number_of_topics)
            if iteration > 100 and abs(current_likelihood - tmp_likelihood) < epsilon/10:
                return tmp_likelihood
            current_likelihood = tmp_likelihood

def get_topics(file_path = 'data/tweet_corpus.txt', top_N = 7, depth = 30):
    documents_path = file_path
    corpus = Corpus(documents_path)  # instantiate corpus
    lines = open(file_path).readlines()
    random.shuffle(lines)
    open(file_path, 'w').writelines(lines)
    corpus.build_corpus()
    corpus.build_vocabulary()
    number_of_topics = 2
    max_iterations = depth
    epsilon = 0.001
    corpus.plsa(number_of_topics, max_iterations, epsilon)
    prob_sums = corpus.topic_word_prob.sum(axis=0)
    top_indices = heapq.nlargest(top_N, range(len(prob_sums)), key=list(prob_sums).__getitem__)
    top_N_topics = []
    for index in top_indices:
        top_N_topics.append(list(corpus.vocabulary)[index])
    return top_N_topics


api/api.py
from flask import Flask, request, render_template
import json
import TwitterAccess

# template_folder is relative to where the main flask run .py file is (/api/app.py)
app = Flask(__name__)# , template_folder='../templates', static_url_path="/")


nytimes_user_id = 2244994945
fox_news_user_id = 1367531
guardian_user_id = 788524
msnbc_user_id = 2836421
politico_user_id = 9300262
abc_news_user_id = 28785486
cbs_news_user_id = 15012486
vice_news_user_id = 1630896181
news_source_ids = [nytimes_user_id,
    fox_news_user_id, 
    guardian_user_id, 
    msnbc_user_id,
    politico_user_id, 
    abc_news_user_id, 
    cbs_news_user_id, 
    vice_news_user_id]


@app.route('/')
def index():
    return render_template("index.html")


@app.route('/query', methods=['POST'])
def query():
    user_query = request.form['query_input']
    return render_template("results.html")


@app.route('/tweets', methods=['GET', 'POST'])
def tweets():
    twitter_client = TwitterAccess.TwitterClient()
    if request.method == 'POST':
        return login_user()
    elif request.method == 'GET':
        '''Getting the query results from Twitter and returning it to the api caller'''
        query = json.loads(request.data)["query"]
        tweets = twitter_client.get_tweets(query=query, count=100)
        positive_tweets = [tweet for tweet in tweets if tweet['sentiment'] == 'positive']
        negative_tweets = [tweet for tweet in tweets if tweet['sentiment'] == 'negative']
        all_tweets = {
            "positive": positive_tweets,
            "negative": negative_tweets
        }
        return tweets

@app.route('/tweets/news', methods=['GET', 'POST'])
def news_tweets():
    query = json.loads(request.data)["query"]
    print(query)
    api = TwitterAccess.TwitterClient()
    if request.method == 'POST':
        '''Getting the query results from Twitter and returning it to the api caller'''
        tweets = api.get_tweets_by_ids(query=query, count=50, id_list=news_source_ids)
        return tweets



@app.route('/user/<username>')
def show_user_profile(username=None):
    # username=None ensures the code run even when no name is provided
    return render_template('user-profile.html', username=username)


@app.route('/post/<int:post_id>')
def show_post(post_id):
    return str(post_id)


def login_user():
    return "You are logged in"


def serve_login_page():
    return "Please log in"


sources.py
nytimes_user_id = 2244994945
fox_news_user_id = 1367531
guardian_user_id = 788524
msnbc_user_id = 2836421
politico_user_id = 9300262
abc_news_user_id = 28785486
cbs_news_user_id = 15012486
vice_news_user_id = 1630896181
news_source_ids = [nytimes_user_id,
    fox_news_user_id, 
    guardian_user_id, 
    msnbc_user_id,
    politico_user_id, 
    abc_news_user_id, 
    cbs_news_user_id, 
    vice_news_user_id]

    
news_sources = [{
    "name": "New York Times",
    "id": nytimes_user_id,
    "username": "nytimes"
},{
    "name": "Fox News",
    "id": fox_news_user_id,
    "username": "foxnews"
},{
    "name": "Guardian News",
    "id": guardian_user_id,
    "username": "guardiannews"
},{
    "name": "MSNBC News",
    "id": msnbc_user_id,
    "username": "msnbc"
},{
    "name": "POLITICO",
    "id": politico_user_id,
    "username": "POLITICO"
},{
    "name": "ABC News",
    "id": abc_news_user_id,
    "username": "abc"
},{
    "name": "CBS News",
    "id": cbs_news_user_id,
    "username": "cbsnews"
},{
    "name": "Vice News",
    "id": vice_news_user_id,
    "username": "vicenews"
}]

 # Names of check box HTML boxes
news_source_names = ["nytimes", "FoxNews", "Guardian", "MSNBC", "Politico", "abcnews", "CBSNews", "VICENews"]

LnRanker.py
# import math
# import sys
# import time

# import metapy
# import pytoml
# import math
# # from scipy import stats

# class InL2Ranker(metapy.index.RankingFunction):
#     """
#     Create a new ranking function in Python that can be used in MeTA.
#     """
#     def __init__(self, some_param=1.0):
#         self.param = some_param
#         # You *must* call the base class constructor here!
#         super(InL2Ranker, self).__init__()

#     def score_one(self, sd):
#         """
#         You need to override this function to return a score for a single term.
#         For fields available in the score_data sd object,
#         @see https://meta-toolkit.org/doxygen/structmeta_1_1index_1_1score__data.html
#         """
#         normalized_term_frequency = sd.doc_term_count * math.log((1+(sd.avg_dl/sd.doc_size)), 2)
#         # print("{}".format(normalized_term_frequency))
#         # old_score = (self.param + sd.doc_term_count) / (self.param * sd.doc_unique_terms + sd.doc_size)
#         # print("{}".format(old_score))
#         score = sd.query_term_weight * (normalized_term_frequency / (normalized_term_frequency + self.param)) * math.log(((sd.num_docs + 1) / (sd.corpus_term_count + 0.5)), 2)
#         # print("{}".format(score))
#         return score

# def load_ranker(cfg_file):
#     """
#     Use this function to return the Ranker object to evaluate, e.g. return InL2Ranker(some_param=1.0)
#     The parameter to this function, cfg_file, is the path to a
#     configuration file used to load the index. You can ignore this for MP2.
#     """
#     return InL2Ranker(some_param=10.0)

# if __name__ == '__main__':
#     if len(sys.argv) != 2:
#         print("Usage: {} config.toml".format(sys.argv[0]))
#         sys.exit(1)

#     cfg = sys.argv[1]
#     print('Building or loading index...')
#     idx = metapy.index.make_inverted_index(cfg)
#     # # Examine number of documents
#     # print("number of docs: {}".format(idx.num_docs()))
#     # # Number of unique terms in the dataset
#     # print("number of unique terms: {}".format(idx.unique_terms()))
#     # # The average document length
#     # print("avg document length: {}".format(idx.avg_doc_length()))
#     # # The total number of terms
#     # print("total corpus terms: {}".format(idx.total_corpus_terms()))

#     inl2_results = []
#     dirichletprior_results = []
#     dirichlet_ranker = metapy.index.DirichletPrior(10)
#     ranker = load_ranker(cfg)
#     ev = metapy.index.IREval(cfg)
#     # dirichev = metapy.index.IREval(cfg)

#     with open(cfg, 'r') as fin:
#         cfg_d = pytoml.load(fin)

#     query_cfg = cfg_d['query-runner']
#     if query_cfg is None:
#         print("query-runner table needed in {}".format(cfg))
#         sys.exit(1)

#     start_time = time.time()
#     top_k = 10
#     query_path = query_cfg.get('query-path', 'queries.txt')
#     query_start = query_cfg.get('query-id-start', 0)


#     query = metapy.index.Document()
#     print('Running queries')
#     with open(query_path) as query_file:
#         for query_num, line in enumerate(query_file):
#             query.content(line.strip())
#             results = ranker.score(idx, query, top_k)
#             # dirichlet_result = dirichlet_ranker.score(idx, query, top_k)
#             avg_p = ev.avg_p(results, query_start + query_num, top_k)
#             # avg_dirich_p = dirichev.avg_p(dirichlet_result, query_start + query_num, top_k)
#             inl2_results.append(avg_p)
#             # dirichletprior_results.append(avg_dirich_p)
#             # print("Query {} average precision: {}".format(query_num + 1, avg_p))
#     # print(inl2_results)
#     # print(dirichletprior_results)
#     print("inl2 mean average precision: {}".format(ev.map()))
#     # print("dirich mean average precision: {}".format(dirichev.map()))
#     print("Elapsed: {} seconds".format(round(time.time() - start_time, 4)))
#     # t_test_p_value = stats.ttest_rel(inl2_results, dirichletprior_results).pvalue
#     # print("T-Test result P value: {}".format(t_test_p_value))
#     # f = open("significance.txt", "w")
#     # f.write(format(t_test_p_value))
#     # f.close()


api/tweets.py
import os
import tweepy

client = tweepy.Client(bearer_token=os.getenv('bearer_token'))

#search by word match
# response = client.search_recent_tweets("biden")

#search by user id received from https://api.twitter.com/2/users/by/username/:username (e.g., nytimes = 807095, FoxNews = 1367531)
nytimes_user_id = 2244994945
fox_news_user_id = 1367531
nytimes_response = client.get_users_tweets(nytimes_user_id)
fox_news_response = client.get_users_tweets(fox_news_user_id)


# The method returns a Response object, a named tuple with data, includes,
# errors, and meta fields
# print(response.meta)

# In this case, the data field of the Response returned is a list of Tweet
# objects
nytimes_tweets = nytimes_response.data
fox_news_tweets = fox_news_response.data

tweets = nytimes_tweets + fox_news_tweets

# Each Tweet object has default ID and text fields
for tweet in tweets:
    print(tweet.id)
    print(tweet.text)

# By default, this endpoint/method returns 10 results
# You can retrieve up to 100 Tweets by specifying max_results
# response = client.search_recent_tweets("Tweepy", max_results=100)

app.py
from flask import Flask, request, render_template, redirect, url_for
from gensim.parsing.preprocessing import remove_stopwords, strip_punctuation, strip_short, strip_non_alphanum
import json
import re
import TwitterAccess
from Plsa import *
from sources import *
# template_folder is relative to where the main flask run .py file is (/api/app.py)
app = Flask(__name__)  # , template_folder='../templates', static_url_path="/")

@app.route('/')
def index():
    recommended_topics = []
    # Call our Topic Modeling function using recent search topics as input to display recommended searches
    recommended_topics = get_topics()
    return render_template("index.html", recommended_search_topics=recommended_topics)

@app.route('/delete_search_topics', methods=['POST'])
def delete_search_topics():
    min_documents_in_corpus = 31

    with open('data/tweet_corpus.txt', 'r', encoding='utf-8') as tc:
        corpus = tc.readlines()
        corpus_len = len(corpus)
        print('Total lines: ', corpus_len)

        if corpus_len > min_documents_in_corpus:
            # Delete the oldest documents from the top of the corpus, while always leaving 30 of the most recent
            # documents (bottom of corpus)
            lines_to_delete = list(range(corpus_len - min_documents_in_corpus))
            print(f"Deleting lines {lines_to_delete[0]} to {lines_to_delete[-1]} in corpus...")

            with open('data/tweet_corpus.txt', 'w', encoding='utf-8') as tcw:
                lines_kept = []
                for number, line in enumerate(corpus):
                    if number not in lines_to_delete:
                        lines_kept.append(number)
                        tcw.write(line)

                print(f"Kept lines {lines_kept[0]} to {lines_kept[-1]}")
        else:
            print(f"Corpus has the minimum amount ({min_documents_in_corpus}) of documents in it already. Can't delete "
                  f"any more documents.")
    return render_template("index.html")

@app.route('/query', methods=['POST'])
def query():
    user_query = request.form['query_input']
    print(f"USER QUERY: {user_query}")
    # Hold the selected news sources to explicitly search if user selected any.
    list_of_news_accounts_to_search = []
    # Hold the number of sentiments (positive, neutral, negative) for each news source
    news_source_sentiments = {}
    # Hold the subjectivity scores for each news source
    news_source_subjectivity = {}
    # Go through each HTML checkbox and append the ones that have been selected, if any, to the
    #  list_of_news_accounts_to_search list.
    for news_source_name in news_source_names:
        if request.form.get(news_source_name) is not None:
            list_of_news_accounts_to_search.append(news_source_name)
    if list_of_news_accounts_to_search:
        print(f"Only searching the following News source Twitter account(s): {list_of_news_accounts_to_search}")
        tweets_to_display = news_tweets(user_query=user_query, news_sources=list_of_news_accounts_to_search)
        # Calculate the sentiments total by news source to display them.
        for news_account in list_of_news_accounts_to_search:
            news_source_sentiments[news_account] = [{'positive': 0,
                                                     'neutral': 0,
                                                     'negative': 0}, 0]

            news_source_subjectivity[news_account] = []

        for tweet in tweets_to_display:
            news_account = tweet['source']

            tweet_sentiment = tweet['sentiment']
            news_source_sentiments[news_account][0][tweet_sentiment] += 1
            news_source_sentiments[news_account][1] += 1

            tweet_subjectivity = tweet['subjectivity']
            news_source_subjectivity[news_account].append(tweet_subjectivity)

        # Calculate the percentage of sentiments for each news account
        for account in news_source_sentiments.keys():
            positive = news_source_sentiments[account][0]['positive']
            neutral = news_source_sentiments[account][0]['neutral']
            negative = news_source_sentiments[account][0]['negative']
            total = positive + neutral + negative
            if total > 0:
                positive_percent = positive / total
                neutral_percent = neutral / total
                negative_percent = negative / total
                news_source_sentiments[account][0]['positive'] = round(positive_percent * 100, 2)
                news_source_sentiments[account][0]['neutral'] = round(neutral_percent * 100, 2)
                news_source_sentiments[account][0]['negative'] = round(negative_percent * 100, 2)

        # Calculate the average subjectivity for each news account
        for account in news_source_subjectivity.keys():
            average_subjectivity = round(sum(news_source_subjectivity[account]) / len(news_source_subjectivity[account]), 2)
            news_source_subjectivity[account] = [average_subjectivity]

    else:
        print(f"No explicit news source(s) selected, so searching all of twitter...")
        tweets_to_display = tweets(user_query=user_query)
    return render_template("results.html", tweet_list=tweets_to_display, query=user_query,
                           news_sentiments=news_source_sentiments, news_subjectivity=news_source_subjectivity)

@app.route('/upvote')
def upvote_post():
    # TODO: Enter logic to take the tweet and put it in user DB
    return "nothing"

@app.route('/tweets', methods=['GET', 'POST'])
def tweets(user_query):
    api = TwitterAccess.TwitterClient()
    if request.method == 'POST':
        '''Getting the query results from Twitter and returning it to the api caller'''
        tweets = api.get_tweets(query=user_query, count=50)
    tweet_corpus = open('data/tweet_corpus.txt', 'a', encoding='utf-8')
    if tweets:
        clean_tweets = document_refinery(api, tweets)
        for clean_tweet in clean_tweets:
            if clean_tweet:
                add_to_corpus(clean_tweet)
        for tweet in tweets:
            if re.search(r"http(s)?://t\.co/([A-Za-z]+[A-Za-z0-9-_]+)", tweet['text']):
                tweet['text'] = re.sub("http(s)?://t\.co/([A-Za-z]+[A-Za-z0-9-_]+)", "", tweet['text'])
    return tweets

@app.route('/tweets/news', methods=['GET', 'POST'])
def news_tweets(user_query=None, news_sources=[]):
    if user_query:
        query = user_query
    else:
        query = request.form['query_input']
    api = TwitterAccess.TwitterClient()
    tweets = []
    if request.method == 'POST':
        '''Getting the query results from Twitter and returning it to the api caller'''
        tweets = api.query_twitter_users(query=query, count=50, user_list=news_sources)
    if tweets:
        clean_tweets = document_refinery(api, tweets)
        for clean_tweet in clean_tweets:
            if clean_tweet:
                add_to_corpus(clean_tweet)
        for tweet in tweets:
            if re.search(r"http(s)?://t\.co/([A-Za-z]+[A-Za-z0-9-_]+)", tweet['text']):
                tweet['text'] = re.sub("http(s)?://t\.co/([A-Za-z]+[A-Za-z0-9-_]+)", "", tweet['text'])
        clean_tweets = document_refinery(api, tweets)
        for clean_tweet in clean_tweets:
            if clean_tweet:
                add_to_corpus(clean_tweet)
    return tweets

@app.route('/tweets/init', methods=['GET', 'POST'])
def tweets_init():
    api = TwitterAccess.TwitterClient()
    tweets = []
    '''Getting the recent tweets from out news sources'''
    tweets = api.query_twitter_users(query=None, count=50, user_list=news_sources)
    return tweets

@app.route('/tweets/topics', methods=['GET', 'POST'])
def get_toptics():
    trends = []
    '''Getting trends based on recent tweets and recent queries'''
    trends = get_topics()
    return {"trends": trends}

@app.route('/user/<username>')
def show_user_profile(username=None):
    # username=None ensures the code run even when no name is provided
    return render_template('user-profile.html', username=username)

@app.route('/post/<int:post_id>')
def show_post(post_id):
    return str(post_id)

def document_refinery(api, tweets):
    """
    Takes in a list of returned tweets from the Twitter api, cleans, adn returns them.
    Tweets are disti
    :param tweets: list of dictionary tweets
    :return: list of clean tweets
    """
    corpus_docs = []
   
    for tweet in tweets:
        clean_tweet = tweet['text'].lower()
        if re.search(r"http(s)?://t\.co/([A-Za-z]+[A-Za-z0-9-_]+)", clean_tweet):
            tweet['text'] = re.sub("http(s)?://t\.co/([A-Za-z]+[A-Za-z0-9-_]+)", "", tweet['text'])
            clean_tweet = re.sub("http(s)?://t\.co/([A-Za-z]+[A-Za-z0-9-_]+)", "", clean_tweet)
        if re.search(r"(?<=^|(?<=[^a-zA-Z0-9-_\.]))@([A-Za-z]+[A-Za-z0-9-_]+)", clean_tweet):
            clean_tweet = re.sub("(?<=^|(?<=[^a-zA-Z0-9-_\.]))@([A-Za-z]+[A-Za-z0-9-_]+)", "", clean_tweet)
        clean_tweet = strip_non_alphanum(strip_punctuation(remove_stopwords(clean_tweet)))
        searches = api.get_recent_searches()
        search_parts = []
        if len(searches) > 2:
            search_parts = get_search_parts(searches)
        if clean_tweet != "":
            for tweet_word in clean_tweet.split():
                if tweet_word in search_parts:
                    clean_tweet = clean_tweet.replace(tweet_word, "").replace("  ", " ")                
            #cleaning again because sometimes after removing recently searched words, it can leave a trailing 's' behind
            clean_tweet = strip_short(strip_non_alphanum(strip_punctuation(remove_stopwords(clean_tweet))), minsize=4)
            # print("This is the cleaned tweet after checking for searched terms: ", clean_tweet)
        if clean_tweet:
            corpus_docs.append(clean_tweet)
            clean_tweet = ""
    return corpus_docs            

def add_to_corpus(corpus_doc):
    document_corpus = open('data/tweet_corpus.txt', 'a', encoding='utf-8')
    document_corpus.write("{}\n".format(corpus_doc.replace("\n", " ")))
    
def get_search_parts(searches):
    temp_parts = []
    for search in searches:
        for word in search.split():
            temp_parts.append(word)
    return temp_parts